masonry/src/tree_arena.rs

```diff
@@ -0,0 +1,297 @@
+        fn find_child<Item>(
+            node: &TreeNode<Item>,
+            id: u64,
+        ) -> Option<(&Item, TreeArenaToken<'_, Item>)> {
+            if node.id == id {
+                return Some((
+                    &node.item,
+                    TreeArenaToken {
+                        children: &node.children,
+                    },
+                ));
+            }
+            for child in &node.children {
+                if let Some((item, token)) = find_child(child, id) {
+                    return Some((item, token));
+                }
+            }
+            None
+        }
+
+        for child in &self.roots {
+            if let Some((item, token)) = find_child(child, id) {
+                return Some((item, token));
+            }
+        }
+
+        None
+    }
+
+    pub fn find_mut(&mut self, id: u64) -> Option<(&mut Item, TreeArenaTokenMut<'_, Item>)> {
+        fn find_child_mut<Item>(
+            node: &mut TreeNode<Item>,
+            id: u64,
+        ) -> Option<(&mut Item, TreeArenaTokenMut<'_, Item>)> {
+            if node.id == id {
+                return Some((
+                    &mut node.item,
+                    TreeArenaTokenMut {
+                        children: &mut node.children,
+                    },
+                ));
+            }
+            for child in &mut node.children {
+                if let Some((item, token)) = find_child_mut(child, id) {
+                    return Some((item, token));
+                }
+            }
+            None
+        }
+
+        for child in &mut self.roots {
+            if let Some((item, token)) = find_child_mut(child, id) {
+                return Some((item, token));
+            }
+        }
+
+        None
+    }
+}
+
+impl<'a, Item> TreeArenaToken<'a, Item> {
+    pub fn has_child(self, id: u64) -> bool {
+        for child in self.children {
+            if child.id == id {
+                return true;
+            }
+        }
+        false
+    }
+
+    pub fn get_child(&self, id: u64) -> Option<(&'_ Item, TreeArenaToken<'_, Item>)> {
+        for child in self.children {
+            if child.id == id {
+                return Some((
+                    &child.item,
+                    TreeArenaToken {
+                        children: &child.children,
+                    },
+                ));
+            }
+        }
+        None
+    }
+
+    pub fn into_child(self, id: u64) -> Option<(&'a Item, TreeArenaToken<'a, Item>)> {
+        for child in &self.children[..] {
+            if child.id == id {
+                return Some((
+                    &child.item,
+                    TreeArenaToken {
+                        children: &child.children,
+                    },
+                ));
+            }
+        }
+        None
+    }
+
+    pub fn iter_children(&self) -> impl Iterator<Item = (&'_ Item, TreeArenaToken<'_, Item>)> {
```

Comment:
This won't be possible in the arena version - every place that uses this should be using `widget.children_ids()`, I think.

Reply:
Good catch. It might be possible depending on the implementation, but for simpler implementations it wouldn't be.